mailpile/plugins/vcard_face.py

```diff
@@ -0,0 +1,57 @@
+import time
+
+from mailpile.i18n import gettext as _
+from mailpile.mailutils import Email
+from mailpile.plugins import PluginManager
+from mailpile.search import MailIndex
+from mailpile.vcard import MailpileVCard, VCardImporter, VCardLine
+
+
+_plugins = PluginManager(builtin=__file__)
+
+
+class FaceImporter(VCardImporter):
+    """
+    Get avatar from Face header.
+    """
+    FORMAT_NAME = 'Face'
+    FORMAT_DESCRIPTION = _('Get avatar from Face header.')
+    SHORT_NAME = 'face'
+    CONFIG_RULES = {
+        'active': [_('Enable this importer'), bool, True],
+    }
+    VCARD_TS = 'x-face-ts'
+    VCARD_IMG = ''
+
+    def get_vcards(self):
+        mail_index = MailIndex(self.session.config)
+        mail_index.load(self.session)
+        results = []
+
+        for vcard in self.session.config.vcards.values():
+            # Get latest message from vcard email address.
+            from_messages = list(mail_index.search(
+                self.session, ['from:' + vcard.email]).as_set())
+            if not from_messages:
+                continue
+            mail_index.sort_results(self.session, from_messages, 'date')
+            latest_message = Email(mail_index, from_messages.pop()).get_msg()
+
+            # Add face vcard.
+            face = mail_index.hdr(latest_message, 'face')
+            if face:
+                results.append(self._make_vcard(vcard.email, VCardLine(
+                    name='photo',
+                    value='data:image/png;base64,%s' % face,
```

Comment:
I don't think this issue is particular to this importer, even though it's more likely to be an invalid image than those from other sources.

We don't need to verify that it's a valid png for security reasons, we just need to escape/sanitize it at some point between parsing and rendering to the client to avoid XSS attacks. This should really happen to all photo's anyway in case one of the avatar services is compromised.

Hopefully that is already taken care of and you're just talking about the user experience issue when a broken image is displayed. The image selection algorithm (currently just the latest imported) needs to get smarter anyway, and I would think "broken image automatically reported by client javascript" would be just another (heavily weighted) factor in the ranking.